util/circular_buffer.c

```diff
@@ -146,12 +141,148 @@ int32_t cb_remove(struct circular_buffer *desc)
  *  - \ref SUCCESS : On success
  *  - \ref FAILURE : Otherwise
  */
-int32_t cb_size(struct circular_buffer *desc, uint32_t *nb_elements)
+int32_t cb_size(struct circular_buffer *desc, uint32_t *size)
+{
+	uint32_t overflows;
+	uint32_t overflowed_size;
+
+	if (!desc || !size)
+		return -EINVAL;
+
+	if (desc->write.nb_overflows < desc->read.nb_overflows)
+		overflows = UINT32_MAX - desc->read.nb_overflows +
+			    desc->write.nb_overflows;
+	else
+		overflows = desc->write.nb_overflows - desc->read.nb_overflows;
+
+	/* Min function is used to avoid an integer overflow. */
+	overflowed_size = desc->size * min(2, overflows);
+
+	*size = overflowed_size + desc->write.idx - desc->read.idx;
+	if (*size > desc->size) {
+		*size = desc->size;
+		return -EOVERRUN;
+	}
+
+	return SUCCESS;
+}
+
+static int32_t cb_prepare_async_operation(struct circular_buffer *desc,
+		uint32_t requested_size,
+		void **buff,
+		uint32_t *raw_size_available,
+		bool is_read)
+{
+	struct cb_ptr	*ptr;
+	uint32_t	available_size;
+	uint32_t	ret;
+
+	if (!desc || !buff || !raw_size_available)
+		return -EINVAL;
+
+	ret = SUCCESS;
+	/* Select if read or write index will be updated */
+	ptr = is_read ? &desc->read : &desc->write;
+
+	/* Only one transaction type possible at a single time */
+	if (ptr->async_started)
```

Comment:
Probably testing and setting of `async_started`, should be made in a critical section.
https://www.nongnu.org/lwip/2_0_x/group__sys__prot.html

Reply:
Added comment in cb_init for this in #694